core/src/main/java/org/mule/runtime/core/DefaultMuleMessage.java

```diff
@@ -1549,14 +1225,14 @@ private MuleMessage copyToInbound(DefaultMuleMessage currentMessage, Object payl
             newInboundProperties.put(name, currentMessage.getOutboundProperty(name));
         }
 
-        newMessage.properties.clearProperties(INBOUND);
-        newMessage.properties.clearProperties(OUTBOUND);
+        newMessage.properties.clearInboundProperties();
+        newMessage.properties.clearOutboundProperties();
 
         for (Map.Entry<String, Serializable> s : newInboundProperties.entrySet())
         {
             DataType<?> propertyDataType = currentMessage.getOutboundPropertyDataType(s.getKey());
 
-            newMessage.setProperty(s.getKey(), s.getValue(), INBOUND, propertyDataType);
+            newMessage.setInboundProperty(s.getKey(), s.getValue(), (DataType<Serializable>) propertyDataType);
```

Comment:
Can propertyDataType be defined as DataType<? extends Serializable> to avoid this cast?